scipy/interpolate/_inverse_distance_weighted.py

```diff
@@ -0,0 +1,298 @@
+        relevant_values[valid_mask] = self.values[i[valid_mask]]
+        np.multiply(weights, relevant_values, where=valid_mask, out=weighted_values)
+
+        # sum(x*w)/sum(w)
+        numerator = np.sum(weighted_values, axis=-1)
+        denominator = np.sum(weights, where=valid_mask, axis=-1)
+
+        # only set values for those points that had real data points within
+        # distance_upper_bound
+        interp_values[to_interpolate_mask] = (
+            numerator[to_interpolate_mask] / denominator[to_interpolate_mask]
+        )
+
+        # handle case where there are any/mutiple equidistant points
+
+        zero_dist_mask_points = np.any(dist == 0, axis=1)
+        if True in zero_dist_mask_points:
+            zero_dist_mask_full = dist == 0
+            values = np.mean(self.values[i], where=zero_dist_mask_full, axis=1)
+            # clunky handle of case when numpy still wants to take the mean for points
+            # with all False (i.e. no points with 0 dists)
+            interp_values[zero_dist_mask_points] = values[~np.isnan(values)]
+
+        new_shape = (
+            original_shape[:-1] + self.values.shape[1:]
+            if self.values.ndim > 1
+            else original_shape[:-1]
+        )
+        interp_values = interp_values.reshape(new_shape)
+
+        return interp_values
+
+    def _global_idw_interpolation(self, xi, weight_func=None, p=2, eps=1e-7):
+        """
+        Perform global Inverse Distance Weighting interpolation.
+
+        In global IDW, all points in the dataset contribute to the interpolation of each
+        query point.
+
+        Parameters
+        ----------
+        xi : ndarray
+            The query points where interpolation is performed.
+        p : float
+            The power parameter of IDW.
+        eps : float, optional
+            Small constant to prevent division by zero in weight calculation.
+
+        Returns
+        -------
+        ndarray
+            The interpolated values at the query points.
+        """
+        xi_flat = xi.reshape(-1, xi.shape[-1])
+        original_shape = xi.shape
+        flattened_shape = xi_flat.shape
+
+        interp_shape = (
+            flattened_shape[:-1] + self.values.shape[1:]
+            if self.values.ndim > 1
+            else flattened_shape[:-1]
+        )
+
+        if np.issubdtype(self.values.dtype, np.complexfloating):
+            interp_values = np.full(
+                interp_shape, self._fill_value, dtype=self.values.dtype
+            )
+        else:
+            interp_values = np.full(interp_shape, self._fill_value)
+
+        for i in range(len(xi_flat)):
+            dist = distance.cdist(xi_flat[i, np.newaxis], self.points).ravel()
+
+            if weight_func is not None:
+                weights = weight_func(dist, p)
+            else:
+                weights = 1.0 / np.maximum(dist, eps) ** p
+
+            zero_dist_mask = dist == 0
+
+            if True in zero_dist_mask:
```

Comment:
Something like `np.any(zero_dist_mask)` moves this loop into NumPy and makes it faster.